lib/src/revset.rs

```diff
@@ -2150,45 +2151,94 @@ impl PartialSymbolResolver for GitRefResolver {
 const DEFAULT_RESOLVERS: &[&'static dyn PartialSymbolResolver] =
     &[&TagResolver, &BranchResolver, &GitRefResolver];
 
-pub type PrefixResolver<'a, T> = Box<dyn Fn(&dyn Repo, &HexPrefix) -> PrefixResolution<T> + 'a>;
+#[derive(Default)]
+struct CommitPrefixResolver<'a> {
+    context: Option<&'a IdPrefixContext>,
+}
+
+impl PartialSymbolResolver for CommitPrefixResolver<'_> {
+    fn resolve_symbol(
+        &self,
+        repo: &dyn Repo,
+        symbol: &str,
+    ) -> Result<Option<Vec<CommitId>>, RevsetResolutionError> {
+        if let Some(prefix) = HexPrefix::new(symbol) {
+            let resolution = self
+                .context
+                .as_ref()
+                .map(|ctx| ctx.resolve_commit_prefix(repo, &prefix))
+                .unwrap_or_else(|| repo.index().resolve_commit_id_prefix(&prefix));
```

Comment:
nit: as a follow up, maybe, we can move the `repo.index().resolve_*()` fallback out of the `IdPrefixContext`.

(no action needed)

Reply:
I don't think you can unless you eliminate `IdPrefixContext::default()`, which has no disambiguation data.

Reply:
I mean `IdPrefixContext`-based resolution could be stacked on top.
```
if Some(r) = context.map(|ctx| ctx.resolve_commit_prefix(repo, &prefix)) {
    return r; // resolve only in DisambiguationData
}
repo.index().resolve(...)
```

This also implies that `shortest_()` methods will have to be proxied through `DefaultSymbolResolver`. I think that's unavoidable to address the shortest length issue. #2416